sync/src/main/java/tech/pegasys/artemis/sync/PeerSync.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.sync;
+
+import static tech.pegasys.artemis.datastructures.networking.libp2p.rpc.GoodbyeMessage.REASON_FAULT_ERROR;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.compute_start_slot_at_epoch;
+import static tech.pegasys.artemis.util.config.Constants.MAX_BLOCK_BY_RANGE_REQUEST_SIZE;
+
+import com.google.common.base.Throwables;
+import com.google.common.primitives.UnsignedLong;
+import java.util.concurrent.CompletableFuture;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
+import tech.pegasys.artemis.networking.eth2.peers.Eth2Peer;
+import tech.pegasys.artemis.networking.eth2.rpc.core.InvalidResponseException;
+import tech.pegasys.artemis.statetransition.BlockImporter;
+import tech.pegasys.artemis.statetransition.StateTransitionException;
+import tech.pegasys.artemis.storage.ChainStorageClient;
+
+public class PeerSync {
+
+  private static final UnsignedLong STEP = UnsignedLong.ONE;
+
+  private Eth2Peer peer;
+  private final UnsignedLong advertisedFinalizedEpoch;
+  private final UnsignedLong advertisedHeadBlockSlot;
+  private final Bytes32 advertisedHeadBlockRoot;
+  private final ChainStorageClient storageClient;
+  private final BlockImporter blockImporter;
+
+  private UnsignedLong latestRequestedSlot;
+  private final CompletableFuture<PeerSyncResult> finalResult = new CompletableFuture<>();
```

Comment:
It's a bit of an anti-pattern to have a "finalResult" type future like this. Ideally we'd just chain `CompletableFuture` calls all the way down and just return that without having to "copy" the result to this future.  The benefit of using full chaining is you guarantee never to "lose" an exception. If an exception throws anywhere in the process it will automatically be passed back up the chain.  Whereas when copying the result if an exception is thrown in the `whenComplete` function within `executeSync` it will be silently ignored because the return value from `requestSyncBlocks` is never used.

There are times when copying to a result future is required but in this case I think we can manage it. See https://github.com/PegaSysEng/artemis/compare/master...ajsutton:chain-futures  We probably should talk about how we want to approach `CompletableFuture` handling - it can easily get unwieldy but since we're just starting to use them if we think about it a bit we can probably define some good patterns and utility methods to support them so we get it right.

Reply:
Agreed. We should talk about how we want to approach CompletableFutures.